### PR TITLE
[Model] Add exact SANA-Video RMSNorm fast path

### PR DESCRIPTION
**Pull request by liuyao0322**
## Purpose

SANA-Video's low-precision RMSNorm has an important numerical contract: after computing the FP32 variance and normalization, it rounds the normalized activation to the parameter dtype before multiplying by the learned weight. Generic fused RMSNorm kernels normally multiply the weight in FP32 and round only once, so replacing this expression directly changes BF16 output bits.

This PR adds a SANA-specific, bit-exact CUDA fast path that:

- preserves the existing `aten::mean` reduction and `rsqrt` behavior;
- fuses BF16-to-FP32 conversion with squaring before the reduction;
- fuses normalization, the explicit BF16 rounding boundary, and the BF16 affine multiply after the reduction;
- verifies the first eligible `(device, dtype, rows, hidden_size)` signature against the original expression using raw BF16 storage equality, and permanently falls back for that signature on a mismatch or launch failure;
- limits the optimization to contiguous NVIDIA CUDA BF16 inference tensors with a contiguous BF16 weight, no bias, no autograd, and at least 2,000,000 elements;
- retains the original PyTorch expression for text-sized tensors, unsupported dtypes/devices/layouts, training, and compiled regions.

The model API, parameter layout, state dict, CPU behavior, and non-BF16 behavior are unchanged. The custom fast path is an eager/CUDA-graph optimization; full-model Inductor compilation intentionally retains the pre-existing PyTorch expression.

## Test Plan

Hardware/software used for CUDA validation:

- NVIDIA H20-3e
- Python 3.12.13
- PyTorch 2.13.0+cu132 / CUDA 13.2
- Triton 3.7.1
- Diffusers 0.40.0 / Transformers 5.14.1

CUDA correctness and integration tests:

```bash
python -m pytest -q \
  tests/diffusion/models/sana_video/test_sana_rms_norm_cuda.py \
  tests/diffusion/models/sana_video/test_transformer_sana_video.py
```

The tests cover all 65,536 BF16 input bit patterns for the square kernel, affine-tail special values, production and masked-tail shapes, first-sight verification, mismatch and launch-error fallback, verified steady state, a real CUDA graph capture/replay, unsupported input fallback, module routing, and the compile fallback.

Static checks:

```bash
ruff check vllm_omni/diffusion/layers/sana_rms_norm.py \
  vllm_omni/diffusion/models/sana_video/transformer_sana_video.py \
  tests/diffusion/models/sana_video/test_sana_rms_norm_cuda.py
ruff format --check vllm_omni/diffusion/layers/sana_rms_norm.py \
  vllm_omni/diffusion/models/sana_video/transformer_sana_video.py \
  tests/diffusion/models/sana_video/test_sana_rms_norm_cuda.py
python tools/pre_commit/check_torch_cuda.py
python tools/pre_commit/check_forbidden_imports.py
python tools/pre_commit/check_spdx_header.py
python tools/pre_commit/check_test_marks.py
python tools/pre_commit/mypy.py 3.10 <changed Python files>
git diff --check
```

Operator benchmarking used 20 warmups and 32 balanced CUDA-Event samples per arm. End-to-end screening used one engine and switched only `SanaRMSNorm` between the original expression and this fast path; it ran on the pending SANA optimization stack to check interoperability with regional GLUMB compilation, fused RoPE, and residual-gate updates.

**vLLM Version:** 0.28.0

**vLLM-Omni Commit:** `beafe5d913b106414e0e42257b35e28aba68b46e`

## Test Result

- CUDA pytest: **89 passed**.
- Ruff, repository policy hooks, mypy, and `git diff --check`: **passed**.
- Every tested fast-path output was raw-storage bit-identical to the original SANA expression.

H20 BF16 operator results:

| Shape | Original median | Fast-path median | Speedup | Peak allocated memory |
| --- | ---: | ---: | ---: | ---: |
| `[2, 32760, 2240]` | 2.8203 ms | 1.0574 ms | 2.67x | 2,348 MB -> 1,761 MB |
| `[2, 9680, 2240]` | 0.8627 ms | 0.3508 ms | 2.46x | 694 MB -> 520 MB |

The small `[2, 300, 2240]` text shape remains on the original path because it is launch-bound.

End-to-end qualification used 480x832, 81-frame SANA-Video 2B T2V and I2V. Each 4-step decoded-video arm had five balanced-order samples:

| Task | Metric | Original median (range) | Fast-path median (range) | Change |
| --- | --- | ---: | ---: | ---: |
| T2V | Transformer CUDA | 10266.181 ms (10265.731-10267.839) | 9837.976 ms (9837.543-9840.161) | 4.17% lower / 1.0435x |
| T2V | Client E2E | 17398.860 ms (17397.221-17408.870) | 16971.119 ms (16971.085-16974.110) | 2.46% lower / 1.0252x |
| I2V | Transformer CUDA | 10772.707 ms (10710.620-10775.556) | 10340.676 ms (10273.914-10351.817) | 4.01% lower / 1.0418x |
| I2V | Client E2E | 18008.962 ms (17919.791-18035.959) | 17574.681 ms (17483.291-17598.468) | 2.41% lower / 1.0247x |

All ten decoded videos per task were raw-bit identical across the two arms. The decoded-request peak allocated and reserved memory were unchanged because VAE decode determines the request peak.

A separate 50-step latent-only A/B checked the full denoise chain (one sample per arm):

| Task | Metric | Original | Fast path | Change |
| --- | --- | ---: | ---: | ---: |
| T2V | Transformer CUDA | 128352.173 ms | 122988.993 ms | 4.18% lower / 1.0436x |
| T2V | Client E2E | 128420.078 ms | 123057.968 ms | 4.18% lower / 1.0436x |
| I2V | Transformer CUDA | 133822.876 ms | 128436.828 ms | 4.03% lower / 1.0419x |
| I2V | Client E2E | 133963.165 ms | 128580.135 ms | 4.02% lower / 1.0419x |

Both 50-step latent outputs were bit-identical (`max_abs = mean_abs = 0`, matching raw SHA-256). Peak allocated memory was 0.354 GiB lower in both runs; peak reserved memory was unchanged.

Each transformer step dispatched 81 SANA RMSNorm calls: 60 video-token calls used the verified fast path and 21 text-sized calls retained the original expression. Thus the 4-step and 50-step fast arms recorded 240 and 3,000 fused calls, respectively, while every original arm recorded zero. No RMSNorm signature was disabled. Regional compilation produced the expected 20 compiled GLUMB modules, two Dynamo graphs, and eight Inductor kernels in both T2V and I2V qualification runs.

### CODE REVIEW

**Comment by liuyao0322** [comment]
Author self-review completed.

- Re-read the Diffusers SANA RMSNorm expression and verified the key numerical contract: the FP32-normalized activation is rounded to BF16 before multiplying the BF16 weight. The fast path keeps the original `aten::mean` reduction and `rsqrt`, and only fuses the two surrounding pointwise regions.
- Checked the explicit FP32 multiply and BF16 RNE boundaries, raw-storage first-sight verification, per-signature fail-closed behavior, NVIDIA/BF16/layout/size/autograd guards, text-shape fallback, and full-compile fallback.
- Checked steady-state and graph behavior, including a real CUDA Graph capture/replay. The tests also cover all 65,536 BF16 square inputs, affine special values, masked tails, model routing, mismatch fallback, and launch-error fallback; 89 H20 CUDA/model tests passed.
- Rechecked the balanced T2V/I2V qualification: five 4-step samples per arm and separate 50-step latent runs were raw-bit identical. Transformer CUDA time was 4.01-4.17% lower, decoded 4-step E2E was 2.41-2.46% lower, and 50-step latent E2E was 4.02-4.18% lower. No signature was disabled.
- The performance scope is single-request H20 BF16 serving; no concurrency or cross-architecture claim is made.

I found no remaining correctness or scope blocker. This is the final PR in the intended review/merge order: **#6647 -> #6819 -> #6823 -> #6825**.

**Comment by liuyao0322** [comment]
@vllm-omni-review-bot, the author self-review is complete. Please assign the appropriate reviewer.

@fhfuih @Bounty-hunter @wtomin @RuixiangMa, this is the final PR in the proposed SANA-Video review/merge order: **#6647 -> #6819 -> #6823 -> #6825**. It can be reviewed independently, but should be merged after #6823.

File: vllm_omni/diffusion/layers/sana_rms_norm.py [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+    if signature not in _VERIFIED_SIGNATURES:
+        if torch.cuda.is_current_stream_capturing():
+            # Synchronizing a first-sight bit comparison is illegal during
+            # graph capture.  A prior eager warmup can verify the signature.
+            return _eager_sana_rms_norm(hidden_states, weight, eps)
+
+    try:
+        output = torch.ops.vllm_omni.exact_sana_rms_norm(hidden_states, weight, eps)
+    except Exception as error:
```
Comment (lishunyang12) [inline vllm_omni/diffusion/layers/sana_rms_norm.py:265]: Re-raise torch.OutOfMemoryError; temporary memory pressure shouldn't permanently disable the fast path.

**Comment by liuyao0322** [comment]
@lishunyang12  Hello, could you please recheck this PR? thank you !
